### ISSUE TITLE
Critical Discrepancy: Pre-training Implementation uses Supervised Labels and Deviates from the paper's implementation

### ISSUE THREAD

**Issue by Mat-thias** [ssykiotis/ELECTRIcity_NILM#4]
I am attempting to reproduce the pre-training results described in the ELECTRIcity paper. I have a misunderstanding between the code implementation and on the paper in terms of pre-training. It paper stated that pre-training was done with the generator trying to reproduce the masked mains power, then this is fed in to the discriminator to classify if it is original or replaced. The generator uses a mse and kdl loss between the original (masked) mains power and the predicted (masked) mains power and the discriminator uses bce loss compared to the mask value.

<img width="760" height="98" alt="Image" src="https://github.com/user-attachments/assets/c26774a2-4f81-4eea-8c05-2df93a93ae3b" />

<img width="806" height="187" alt="Image" src="https://github.com/user-attachments/assets/2e984e24-ec9c-4cea-b2e7-33baf59bd22d" />

But the code implementation does not seem to follow this.

The output of the model (when in pretraining) is a tuple of (disc, gen).
```python

    def forward(self,sequence,mask=None):
        if self.pretrain:
            gen_out = self.Generator(sequence)
            disc_in = sequence
            disc_in[mask] = gen_out[mask]
        else:
            disc_in = sequence
        disc_out = self.Discriminator(disc_in)
        if self.pretrain:
            return disc_out, gen_out
        else:
            return disc_out, None
```

But during training, it is the generator output that is compared to the mask and the discriminator output is compared with the appliance output.
I have attached the relevant part related to pretraining and added inline explanation to follow the flow of the input and output towards the loss.
```python

    def pretrain_one_epoch(self,epoch):
        ....
        for _,batch in enumerate(tqdm_dataloader):
            # from the pretrain_dataloader
            # x is the masked mains_power
            # y is the appliance_power
            # status is the appliance_power
            x, y, status = [batch[i].to(self.device) for i in range(3)]
            mask = (status >= 0)

            y_capped = y / self.cutoff
            # logits is the disc_out
            logits, gen_out, logits_y, logits_status    = self.get_model_outputs(x,mask)

            logits_masked        = torch.masked_select(logits       , mask).view((-1))
            # labels_maksed is the appliance_power
            labels_masked        = torch.masked_select(y_capped     , mask).view((-1))
            gen_out = gen_out.view(-1)

            mask = mask.view(-1).type(torch.DoubleTensor).to(self.device)

            total_loss = self.loss_fn_pretrain(logits_masked,labels_masked,gen_out,mask)
            ...

    def get_model_outputs(self, x,mask=None):
        # x is the mains_power
        logits, gen_out = self.model(x,mask)
        logits_y        = self.cutoff_energy(logits*self.cutoff)
        logits_status   = self.compute_status(logits_y)    
        # output: logits is disc_out
        return logits, gen_out, logits_y, logits_status


    def loss_fn_gen(self,logits_masked,labels_masked):
        mse_arg_1 = logits_masked.contiguous().view(-1).double()
        mse_arg_2 = labels_masked.contiguous().view(-1).double()
        kl_arg_1  = torch.log(F.softmax(logits_masked.squeeze()/self.tau, dim=-1) + 1e-9)
        kl_arg_2  = F.softmax(labels_masked.squeeze()/self.tau, dim=-1)
        mse_loss = self.mse(mse_arg_1,mse_arg_2)
        kl_loss  = self.kl(kl_arg_1,kl_arg_2)
        loss = mse_loss + kl_loss
        return loss
        
    def loss_fn_disc(self,gen_out,mask):
        return self.bceloss(gen_out,mask)

    def loss_fn_pretrain(self,logits_masked,labels_masked,gen_out,mask):

        # The logits_masked (disc_out) is compared with the labels_masked (appliance_power)
        # using kdl and mse loss
        # this contradicts what the paper said, 
        gen_loss  = self.loss_fn_gen(logits_masked,labels_masked)

        # NOTE: The gen_out is compared with mask using the bceloss, not the disc_out
        disc_loss = self.loss_fn_disc(gen_out,mask)
        return gen_loss + disc_loss 
```

The implementation contradicts what was stated and paper. What the code does during the said "pretraining", it masks the mains power going into the generator, this output is compared or sees its target as the logits of the mask, Then passes this logits as a corrupted version of the mains power to the discriminator which is then compared to the appliance power, this basically reduces to the same implementation as bert4nilm, which is the explanation for the negligible difference in the result and that of bert4nilm. The output of the generator is "basically" useless or at best can be thought of just as corrupter (or a different masking approach) for the discriminator.
